vicinity/backends/faiss.py

```diff
@@ -0,0 +1,229 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Literal
+
+import faiss
+import numpy as np
+from numpy import typing as npt
+
+from vicinity.backends.base import AbstractBackend, BaseArgs
+from vicinity.datatypes import Backend, QueryResult
+from vicinity.utils import normalize
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class FaissArgs(BaseArgs):
+    dim: int = 0
+    index_type: Literal["flat", "ivf", "hnsw", "lsh", "scalar", "pq", "ivf_scalar", "ivfpq", "ivfpqr"] = "hnsw"
+    metric: Literal["cosine", "l2"] = "cosine"
+    nlist: int = 100
+    m: int = 8
+    nbits: int = 8
+    refine_nbits: int = 8
+    direct_map: bool = True
+
+
+class FaissBackend(AbstractBackend[FaissArgs]):
+    argument_class = FaissArgs
+
+    def __init__(
+        self,
+        index: faiss.Index,
+        arguments: FaissArgs,
+    ) -> None:
+        """Initialize the backend using a FAISS index."""
+        super().__init__(arguments)
+        self.index = index
+        # Enable DirectMap if specified and supported by index type
+        if isinstance(index, faiss.IndexIVF) and arguments.direct_map:
+            index.set_direct_map_type(faiss.DirectMap.Hashtable)
+
+    @classmethod
+    def from_vectors(  # noqa: C901
+        cls: type[FaissBackend],
+        vectors: npt.NDArray,
+        index_type: Literal["flat", "ivf", "hnsw", "lsh", "scalar", "pq", "ivf_scalar", "ivfpq", "ivfpqr"] = "flat",
+        metric: Literal["cosine", "l2"] = "cosine",
+        nlist: int = 100,
+        m: int = 8,
+        nbits: int = 8,
+        refine_nbits: int = 8,
+        direct_map: bool = True,
+        **kwargs: Any,
+    ) -> FaissBackend:
+        """
+        Create a new instance from vectors.
+
+        :param vectors: The vectors to index.
+        :param index_type: The type of FAISS index to use.
+        :param metric: The metric to use for similarity search.
+        :param nlist: The number of cells for IVF indexes.
+        :param m: The number of subquantizers for PQ and HNSW indexes.
+        :param nbits: The number of bits for LSH and PQ indexes.
+        :param refine_nbits: The number of bits for the refinement stage in IVFPQR indexes.
+        :param direct_map: Whether to enable DirectMap for IVF indexes. This allows deletion of vectors.
+        :param **kwargs: Additional arguments to pass to the backend.
+        :return: A new FaissBackend instance.
+        :raises ValueError: If an invalid index type is provided.
+        """
+        dim = vectors.shape[1]
+
+        # If using cosine, normalize vectors to unit length
+        if metric == "cosine":
+            vectors = normalize(vectors)
+            faiss_metric = faiss.METRIC_INNER_PRODUCT
+        else:
+            faiss_metric = faiss.METRIC_L2
+
+        if index_type == "flat":
```

Comment:
I think this whole tree can be simplified by 

1. recognizing that some indexes need training and others don't. Defining which ones need training and only having one `index.train` line should be sufficient. 

2. you can see if the index_type starts with `ivf`, and then remove all the boilerplate that comes with the rest.